(jdef) this patch allows client.Config.Prefix to be automatically determined by
parsing the KUBERNETES_MASTER environment variable. Super useful for clients of
DCOS, where the apiserver runs behind the admin proxy router that's doing
proxypass style URL translation.

diff --git a/pkg/client/clientcmd/client_config.go b/pkg/client/clientcmd/client_config.go
index 3b29903..616d1cb 100644
--- a/pkg/client/clientcmd/client_config.go
+++ b/pkg/client/clientcmd/client_config.go
@@ -18,6 +18,7 @@ package clientcmd
 
 import (
 	"io"
+	"net/url"
 	"os"
 
 	"github.com/imdario/mergo"
@@ -86,6 +87,18 @@ func (config DirectClientConfig) ClientConfig() (*client.Config, error) {
 
 	clientConfig := &client.Config{}
 	clientConfig.Host = configClusterInfo.Server
+	if u, err := url.ParseRequestURI(clientConfig.Host); err == nil && u.Opaque == "" && len(u.Path) > 1 {
+		if u.Scheme != "" {
+			clientConfig.Host = u.Scheme + "://"
+		} else {
+			clientConfig.Host = "http://"
+		}
+		if u.User != nil {
+			clientConfig.Host += u.User.String() + "@"
+		}
+		clientConfig.Host += u.Host
+		clientConfig.Prefix = u.Path
+	}
 	clientConfig.Version = configClusterInfo.APIVersion
 
 	// only try to read the auth information if we are secure
